qiita_db/analysis.py

```diff
@@ -525,19 +525,70 @@ def can_be_publicized(self):
 
         Returns
         -------
-        bool
-            Whether the analysis can be publicized or not
+        bool, list
+            Whether the analysis can be publicized or not and the list of not
+            public artifacts
         """
         # The analysis can be made public if all the artifacts used
         # to get the samples from are public
         with qdb.sql_connection.TRN:
+            non_public = []
             sql = """SELECT DISTINCT artifact_id
                      FROM qiita.analysis_sample
-                     WHERE analysis_id = %s"""
+                     WHERE analysis_id = %s
+                     ORDER BY artifact_id"""
+            qdb.sql_connection.TRN.add(sql, [self.id])
+            for aid in qdb.sql_connection.TRN.execute_fetchflatten():
+                if qdb.artifact.Artifact(aid).visibility != 'public':
+                    non_public.append(aid)
+
+            return (non_public == [], non_public)
+
+    @property
+    def is_public(self):
+        """Returns if the analysis is public
+
+        Returns
+        -------
+        bool
+            If the analysis is public
+        """
+        with qdb.sql_connection.TRN:
+            # getting all root artifacts / command_id IS NULL
+            sql = """SELECT visibility
```

Comment:
Minor, but this query could probably also use a `DISTINCT`. Though the only noticeable benefits would be if this table is returning a lot of visibilities.